turbine/src/addr_cache.rs

```diff
@@ -0,0 +1,412 @@
+    code: Vec<Option<(/*root_distance:*/ u8, Box<[SocketAddr]>)>>,
+    data: Vec<Option<(/*root_distance:*/ u8, Box<[SocketAddr]>)>>,
+    // Code and data indices where [..index] are fully populated.
+    index_code: usize,
+    index_data: usize,
+    // Maximum code and data indices observed in retransmit-stage.
+    max_index_code: u32,
+    max_index_data: u32,
+    // If the last data shred in the slot is already observed, the cache is no
+    // longer extended beyond max_index_{code,data}.
+    last_shred_in_slot: bool,
+}
+
+impl AddrCache {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            capacity,
+            num_shreds: 0,
+            window: VecDeque::new(),
+            counts: HashMap::new(),
+            cache: {
+                // 2x capacity in order to implement lazy eviction.
+                let capacity = capacity.saturating_mul(2).saturating_add(1);
+                HashMap::with_capacity(capacity)
+            },
+        }
+    }
+
+    // Returns (root-distance, socket-addresses) cached for the given shred-id.
+    #[inline]
+    pub(crate) fn get(&self, shred: &ShredId) -> Option<(/*root_distance:*/ u8, &[SocketAddr])> {
+        self.cache
+            .get(&shred.slot())?
+            .get(shred.shred_type(), shred.index())
+    }
+
+    // Stores (root-distance, socket-addresses) precomputed speculatively for
+    // the given shred-id.
+    pub(crate) fn put(
+        &mut self,
+        shred: &ShredId,
+        entry: (/*root_distance:*/ u8, Box<[SocketAddr]>),
+    ) {
+        self.get_cache_entry_mut(shred.slot())
+            .put(shred.shred_type(), shred.index(), entry);
+        self.maybe_trim_cache();
+    }
+
+    // Records data observed from incoming shreds at retransmit stage.
+    pub(crate) fn record(&mut self, slot: Slot, stats: &mut RetransmitSlotStats) {
+        // All addresses should be for the same slot.
+        debug_assert!(stats.addrs.iter().all(|(shred, _, _)| shred.slot() == slot));
+        // Update rolling window count of shreds per slot.
+        let num_shreds: usize = stats.num_shreds_received.iter().sum();
+        if num_shreds > 0 {
+            self.num_shreds += num_shreds;
+            self.window.push_back((slot, num_shreds));
+            *self.counts.entry(slot).or_default() += num_shreds;
+            self.maybe_trim_slot_counts();
+        }
+        debug_assert!(self.verify());
+        // If there are no addresses to cache and the cache entry is not
+        // allocated for the slot yet, then ignore.
+        if stats.addrs.is_empty() && !self.cache.contains_key(&slot) {
+            return;
+        }
+        // Update the cached entry for the slot.
+        let entry = self.get_cache_entry_mut(slot);
+        entry.max_index_code = entry.max_index_code.max(stats.max_index_code);
+        entry.max_index_data = entry.max_index_data.max(stats.max_index_data);
+        entry.last_shred_in_slot |= stats.last_shred_in_slot;
```

Comment:
Could we have a metric here to report how many data shreds we precomputed vs how many were necessary based on this flag? Gives some insight on any wasted computation for future tuning.

Reply:
sure, but can we do that in follow up PRs?
this code is already ~700 lines.

Also, the `EXTEND_BUFFER` limits how far the cache goes beyond the `max_index_{code,data}`.
With current values it does not go more than 512 indices beyond that.
(and if it has observed `last_shred_in_slot` it stops extending).

Reply:
ah got it, i overlooked the part that it's limited by the max index